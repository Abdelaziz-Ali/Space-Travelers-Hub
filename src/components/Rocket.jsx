import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserve, cancelReservation } from '../redux/rockets/rocketSlice';

export default function Rocket({
  image, description, name, id, isReserved,
}) {
  const className = isReserved ? 'bg-success-subtle' : '';
  return (
    <figure className={`container mb-2 border rounded p-3 ${className}`}>
      <div className="row align-items-center">

        <div className="col-3">
          <img src={image} alt={name} className="img-thumbnail" />
        </div>

        <div className="col-9">
          <h3>{name}</h3>
          <p>
            {isReserved && <Badge />}
            {description}
          </p>
          {isReserved ? <CancelBtn id={id} /> : <ReserveBtn id={id} /> }
        </div>

      </div>
    </figure>
  );
}

Rocket.propTypes = {
  image: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  isReserved: PropTypes.bool.isRequired,
};

function ReserveBtn({ id }) {
  const dispatch = useDispatch();
  return (
    <button type="button" className="btn btn-success" onClick={() => dispatch(reserve(id))}>
      Reserve Rocket
    </button>
  );
}

ReserveBtn.propTypes = {
  id: PropTypes.string.isRequired,
};

function CancelBtn({ id }) {
  const dispatch = useDispatch();
  return (
    <button
      type="button"
      className="btn btn-outline-secondary"
      onClick={() => dispatch(cancelReservation(id))}
    >
      Cancel Reservation
    </button>
  );
}

CancelBtn.propTypes = {
  id: PropTypes.string.isRequired,
};

function Badge() {
  return (
    <>
      <span className="badge text-bg-info">Reserved</span>
      &nbsp;
    </>
  );
}
