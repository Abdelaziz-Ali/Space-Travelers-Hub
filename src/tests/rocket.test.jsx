import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import Rocket from '../components/Rocket';
import store from '../redux/store';

describe('Tests for rocket component', () => {
  const rocket = (
    <Provider store={store}>
      <Rocket description="Rockets" id="1" image="none" isReserved={false} name="roket" />
    </Provider>
  );
  test('Create snapshot for rocket', () => {
    const component = render(rocket);
    expect(component).toMatchSnapshot();
  });

  test('Rocket has a button element', () => {
    render(rocket);
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
  });

  test('It has an image component', () => {
    render(rocket);
    const image = screen.getByRole('img');
    expect(image).toBeInTheDocument();
  });

  test('The reserve roket button is not focused by default', () => {
    render(rocket);
    const button = screen.getByRole('button');
    expect(button).not.toHaveFocus();
  });
});
