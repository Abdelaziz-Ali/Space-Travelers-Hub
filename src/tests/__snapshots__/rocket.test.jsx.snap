// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for rocket component Create snapshot for rocket 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <figure
        class="container mb-2 border rounded p-3 "
      >
        <div
          class="row align-items-center"
        >
          <div
            class="col-3"
          >
            <img
              alt="roket"
              class="img-thumbnail"
              src="none"
            />
          </div>
          <div
            class="col-9"
          >
            <h3>
              roket
            </h3>
            <p>
              Rockets
            </p>
            <button
              class="btn btn-success"
              type="button"
            >
              Reserve Rocket
            </button>
          </div>
        </div>
      </figure>
    </div>
  </body>,
  "container": <div>
    <figure
      class="container mb-2 border rounded p-3 "
    >
      <div
        class="row align-items-center"
      >
        <div
          class="col-3"
        >
          <img
            alt="roket"
            class="img-thumbnail"
            src="none"
          />
        </div>
        <div
          class="col-9"
        >
          <h3>
            roket
          </h3>
          <p>
            Rockets
          </p>
          <button
            class="btn btn-success"
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </figure>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
